File: tests/test_jma_alerts.js
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const vm = require('node:vm');

const stageRoot = path.resolve(__dirname, '..');
const fixtureRoot = path.resolve(__dirname, 'fixtures', 'jma', 'derived');
const rawFixtureRoot = path.resolve(__dirname, 'fixtures', 'jma', 'raw');

function loadJson(name) {
    return JSON.parse(fs.readFileSync(path.join(fixtureRoot, name), 'utf8'));
}

function loadRawJson(name) {
    return JSON.parse(fs.readFileSync(path.join(rawFixtureRoot, name), 'utf8'));
}

function loadUiNormalizer() {
    const source = fs.readFileSync(path.join(stageRoot, 'js', 'ui.js'), 'utf8');
    const context = vm.createContext({ console });
    vm.runInContext(`${source}\nglobalThis.__normalizeJmaAlerts = normalizeJmaAlerts;`, context);
    return context.__normalizeJmaAlerts;
}

function loadWorker() {
    const source = fs
        .readFileSync(path.join(stageRoot, 'cloudflare-push-worker.js'), 'utf8')
        .replace('export default {', 'globalThis.__worker = {');
    const context = vm.createContext({
        console,
        Date,
        JSON,
        Set,
        Map,
        fetch: undefined
    });
    vm.runInContext(source, context);
    return { worker: context.__worker, context };
}

const expectedCases = [
    ['no-warning-katsushika.json', 0, null],
    ['advisory-katsushika.json', 2, 'レベル2 大雨注意報'],
    ['level3-katsushika.json', 3, 'レベル3 大雨警報'],
    ['level4-katsushika.json', 4, 'レベル4 大雨危険警報'],
    ['level5-katsushika.json', 5, 'レベル5 大雨特別警報'],
    ['continuing-katsushika.json', 3, 'レベル3 大雨警報'],
    ['downgraded-katsushika.json', 2, 'レベル2 大雨注意報'],
    ['released-katsushika.json', 0, null]
];

const normalizeUi = loadUiNormalizer();
const { worker, context } = loadWorker();

for (const [file, level, name] of expectedCases) {
    const data = loadJson(file);
    const uiAlerts = normalizeUi(data, '1312200');
    const workerAlerts = worker.normalizeJMAWarnings(data, '1312200').active;

    assert.equal(uiAlerts.length, level === 0 ? 0 : 1, `UI count: ${file}`);
    assert.equal(workerAlerts.length, level === 0 ? 0 : 1, `Worker count: ${file}`);
    if (level > 0) {
        assert.equal(uiAlerts[0].level, level, `UI level: ${file}`);
        assert.equal(workerAlerts[0].level, level, `Worker level: ${file}`);
        assert.equal(uiAlerts[0].name, name, `UI name: ${file}`);
        assert.equal(workerAlerts[0].name, name, `Worker name: ${file}`);
    }
}

assert.throws(
    () => normalizeUi({ areaTypes: [] }, '1312200'),
    /Unexpected JMA warning response schema/
);
assert.throws(
    () => worker.normalizeJMAWarnings({ areaTypes: [] }, '1312200'),
    /Unexpected JMA warning response schema/
);

const rawLevel2 = normalizeUi(
    loadRawJson('archive-r8-130000-2026060218-level2.json'),
    '1312200'
);
assert(rawLevel2.some(alert => alert.id === 'VPWW55:10' && alert.status === '発表'));

const rawLevel3 = normalizeUi(
    loadRawJson('archive-r8-130000-2026060303-level3-level4-continuing.json'),
    '1312200'
);
assert(rawLevel3.some(alert => alert.id === 'VPWW55:03' && alert.status === '継続'));

const rawDowngrade = normalizeUi(
    loadRawJson('archive-r8-130000-2026060306-downgrade.json'),
    '1312200'
);
assert(rawDowngrade.some(
    alert => alert.id === 'VPWW55:10' && alert.status === '警報から注意報'
));

// Worker notification state: new -> continuing -> downgrade -> release -> reissue.
const kvStore = new Map();
const env = {
    KV: {
        async get(key) {
            return kvStore.get(key) ?? null;
        },
        async put(key, value) {
            kvStore.set(key, value);
        }
    }
};
const notifications = [];
worker.sendToAll = async (_env, payload) => {
    notifications.push(payload);
    return { sent: 1, failed: 0, cleaned: 0 };
};

let responseData = loadJson('level3-katsushika.json');
context.fetch = async () => ({
    ok: true,
    status: 200,
    async json() {
        return responseData;
    }
});

(async () => {
    await worker.checkJMAWarnings(env);
    assert.equal(notifications.length, 1, 'new level 3 must notify once');

    responseData = loadJson('continuing-katsushika.json');
    await worker.checkJMAWarnings(env);
    assert.equal(notifications.length, 1, 'continuing warning must not notify again');

    responseData = loadJson('downgraded-katsushika.json');
    await worker.checkJMAWarnings(env);
    assert.equal(notifications.length, 2, 'downgrade must notify once');
    assert.match(notifications[1].title, /切替/);

    responseData = loadJson('released-katsushika.json');
    await worker.checkJMAWarnings(env);
    assert.equal(notifications.length, 2, 'advisory release must not create urgent notification');

    responseData = loadJson('level3-katsushika.json');
    responseData[0].controlDatetime = '2026-06-03T08:31:05Z';
    responseData[0].reportDatetime = '2026-06-03T17:31:00+09:00';
    await worker.checkJMAWarnings(env);
    assert.equal(notifications.length, 3, 'reissue after release must notify');

    console.log(`JavaScript JMA tests passed: ${expectedCases.length} cases + raw fixtures + state transitions`);
})().catch(error => {
    console.error(error);
    process.exitCode = 1;
});

File: tests/test_jma_alerts.py
import importlib.util
import json
import sys
import types
from pathlib import Path


STAGE_ROOT = Path(__file__).resolve().parents[1]
PROJECT_SCRIPTS = STAGE_ROOT / "scripts"
FIXTURE_ROOT = Path(__file__).resolve().parent / "fixtures" / "jma" / "derived"
RAW_FIXTURE_ROOT = Path(__file__).resolve().parent / "fixtures" / "jma" / "raw"

# 正規化関数の単体テストではネットワーク・Gemini依存を使用しない。
requests_stub = types.ModuleType("requests")
dotenv_stub = types.ModuleType("dotenv")
dotenv_stub.load_dotenv = lambda *_args, **_kwargs: None
google_stub = types.ModuleType("google")
genai_stub = types.ModuleType("google.genai")
google_stub.genai = genai_stub
data_analysis_stub = types.ModuleType("data_analysis")
data_analysis_stub.analyze_data_comprehensive = lambda *_args, **_kwargs: {}
sys.modules.update(
    {
        "requests": requests_stub,
        "dotenv": dotenv_stub,
        "google": google_stub,
        "google.genai": genai_stub,
        "data_analysis": data_analysis_stub,
    }
)

sys.path.insert(0, str(PROJECT_SCRIPTS))
spec = importlib.util.spec_from_file_location(
    "staged_ai_advisor", STAGE_ROOT / "scripts" / "ai_advisor.py"
)
module = importlib.util.module_from_spec(spec)
assert spec.loader is not None
spec.loader.exec_module(module)


def load_json(name):
    return json.loads((FIXTURE_ROOT / name).read_text(encoding="utf-8"))


cases = [
    ("no-warning-katsushika.json", 0, None),
    ("advisory-katsushika.json", 2, "レベル2 大雨注意報"),
    ("level3-katsushika.json", 3, "レベル3 大雨警報"),
    ("level4-katsushika.json", 4, "レベル4 大雨危険警報"),
    ("level5-katsushika.json", 5, "レベル5 大雨特別警報"),
    ("continuing-katsushika.json", 3, "レベル3 大雨警報"),
    ("downgraded-katsushika.json", 2, "レベル2 大雨注意報"),
    ("released-katsushika.json", 0, None),
]

for filename, level, name in cases:
    result = module.normalize_jma_alerts(load_json(filename), "1312200")
    expected_count = 0 if level == 0 else 1
    assert len(result["alerts"]) == expected_count, filename
    if level:
        assert result["alerts"][0]["level"] == level, filename
        assert result["alerts"][0]["name"] == name, filename

raw_level2 = module.normalize_jma_alerts(
    json.loads(
        (RAW_FIXTURE_ROOT / "archive-r8-130000-2026060218-level2.json").read_text(
            encoding="utf-8"
        )
    ),
    "1312200",
)
assert any(
    alert["id"] == "VPWW55:10" and alert["status"] == "発表"
    for alert in raw_level2["alerts"]
)

raw_level3 = module.normalize_jma_alerts(
    json.loads(
        (
            RAW_FIXTURE_ROOT
            / "archive-r8-130000-2026060303-level3-level4-continuing.json"
        ).read_text(encoding="utf-8")
    ),
    "1312200",
)
assert any(
    alert["id"] == "VPWW55:03" and alert["status"] == "継続"
    for alert in raw_level3["alerts"]
)

raw_downgrade = module.normalize_jma_alerts(
    json.loads(
        (RAW_FIXTURE_ROOT / "archive-r8-130000-2026060306-downgrade.json").read_text(
            encoding="utf-8"
        )
    ),
    "1312200",
)
assert any(
    alert["id"] == "VPWW55:10" and alert["status"] == "警報から注意報"
    for alert in raw_downgrade["alerts"]
)

try:
    module.normalize_jma_alerts({"areaTypes": []}, "1312200")
except ValueError as error:
    assert "Unexpected JMA warning response schema" in str(error)
else:
    raise AssertionError("legacy schema must be rejected")

print(f"Python JMA tests passed: {len(cases)} cases + raw fixtures")
